core/trace-experimental/src/main/scala-2/org/typelevel/otel4s/trace/experimental/withSpan.scala

```diff
@@ -0,0 +1,247 @@
+            ),
+          args.collectFirst { case q"debug = true" => }.isDefined
+        )
+      case _ =>
+        (None, false)
+    }
+
+    def spanName(definition: ValOrDefDef): Tree =
+      nameParam match {
+        case Some(Left(const)) =>
+          val literal = Literal(const)
+          q"$literal"
+
+        case Some(Right(tree)) =>
+          tree
+
+        case None =>
+          @annotation.tailrec
+          def resolveEnclosingName(symbol: Symbol, output: String): String =
+            if (symbol.isClass) {
+              val className = symbol.name.toString
+              if (className.startsWith("$anon"))
+                resolveEnclosingName(symbol.owner, "$anon" + "." + output)
+              else
+                className + "." + output
+            } else {
+              resolveEnclosingName(symbol.owner, output)
+            }
+
+          val prefix = resolveEnclosingName(c.internal.enclosingOwner, "")
+
+          val literal = Literal(
+            Constant(prefix + definition.name.decodedName.toString)
+          )
+          q"$literal"
+      }
+
+    def expandDef(defDef: DefDef): Tree = {
+      val effectType = resolveEffectType(defDef.tpt)
+      val name: Tree = spanName(defDef)
+
+      val attributes = defDef.vparamss.flatten.flatMap {
+        case ValDef(mods, name, tpt, _) =>
+          mods.annotations.flatMap { annotation =>
+            val typed = c.typecheck(annotation)
+            if (
+              typed.tpe.typeSymbol.fullName == "org.typelevel.otel4s.trace.experimental.spanAttribute"
+            ) {
+              val keyArg = typed match {
+                case q"new ${_}(${keyArg})" =>
+                  keyArg
+                case _ =>
+                  abort("unknown structure of the @spanAttribute annotation.")
+              }
+
+              val key = keyArg match {
+                // the key param is not specified, use param name as a key
+                case Select(_, TermName("$lessinit$greater$default$1")) =>
+                  Literal(Constant(name.decodedName.toString))
+
+                // type of the AttributeKey must match the parameter type
+                case q"$_[$tpe]($_)(..$_)" =>
+                  val keyType = c.untypecheck(tpe)
+                  val argType = c.typecheck(tpt, c.TYPEmode)
+
+                  if (!keyType.equalsStructure(argType)) {
+                    abort(
+                      s"the argument [${name.toString}] type [$argType] does not match the type of the attribute [$keyType]."
+                    )
+                  }
+                  keyArg
+
+                case _ =>
+                  keyArg
+              }
+
+              ensureImplicitExist(
+                q"_root_.org.typelevel.otel4s.AttributeKey.KeySelect[$tpt]",
+                e =>
+                  s"the argument [${name.decodedName}] cannot be used as an attribute. The type [$tpt] is not supported.${e.getMessage}"
+              )
```

Comment:
Since KeySelect is sealed there's no way to define a `spanAttribute` for an unsupported type, normally we'd create an Attribute manually like
```
case class Foo(value: String)
val attribute = Attribute("foo.value", foo.value)
```
This means we can't use the macro at all if we require one attribute that is unsupported. Would it be sensible to expose some way of contramapping an existing KeySelect?

Reply:
Though thinking about it still won't be possible without some other function from value => key type 😢

Reply:
Valid points. The macro wouldn't work with newtypes/type aliases. 

I guess we can provide some utility methods to lift arbitrary types into the Attribute.